pkg/reconciler/pipelinerun/resources/resultrefresolution.go

```diff
@@ -115,58 +115,74 @@ func removeDup(refs ResolvedResultRefs) ResolvedResultRefs {
 // then a nil list and error is returned instead.
 func convertToResultRefs(pipelineRunState PipelineRunState, target *ResolvedPipelineTask) (ResolvedResultRefs, string, error) {
 	var resolvedResultRefs ResolvedResultRefs
-	for _, ref := range v1.PipelineTaskResultRefs(target.PipelineTask) {
-		resolved, pt, err := resolveResultRef(pipelineRunState, ref)
-		if err != nil {
-			return nil, pt, err
+	for _, resultRef := range v1.PipelineTaskResultRefs(target.PipelineTask) {
+		referencedPipelineTask := pipelineRunState.ToMap()[resultRef.PipelineTask]
+		if referencedPipelineTask == nil {
+			return nil, resultRef.PipelineTask, fmt.Errorf("could not find task %q referenced by result", resultRef.PipelineTask)
+		}
+		if !referencedPipelineTask.isSuccessful() && !referencedPipelineTask.isFailure() {
+			return nil, resultRef.PipelineTask, fmt.Errorf("task %q referenced by result was not finished", referencedPipelineTask.PipelineTask.Name)
+		}
+		// Custom Task
+		switch {
+		case referencedPipelineTask.IsCustomTask():
+			resolved, err := resolveCustomResultRef(referencedPipelineTask.CustomRuns, resultRef)
+			if err != nil {
+				return nil, resultRef.PipelineTask, err
+			}
+			resolvedResultRefs = append(resolvedResultRefs, resolved)
+		default:
+			// Matrixed referenced Pipeline Task
+			if len(referencedPipelineTask.TaskRuns) > 1 {
+				arrayValues, err := findResultValuesForMatrix(referencedPipelineTask, resultRef)
+				if err != nil {
+					return nil, resultRef.PipelineTask, err
+				}
+				for _, taskRun := range referencedPipelineTask.TaskRuns {
+					resolved := createMatrixedTaskResultForParam(taskRun.Name, arrayValues, resultRef)
+					resolvedResultRefs = append(resolvedResultRefs, resolved)
+				}
+			} else {
+				// Regular PipelineTask
+				resolved, err := resolveResultRef(referencedPipelineTask.TaskRuns, resultRef)
+				if err != nil {
+					return nil, resultRef.PipelineTask, err
+				}
+				resolvedResultRefs = append(resolvedResultRefs, resolved)
+			}
 		}
-		resolvedResultRefs = append(resolvedResultRefs, resolved)
 	}
 	return resolvedResultRefs, "", nil
 }
 
-func resolveResultRef(pipelineState PipelineRunState, resultRef *v1.ResultRef) (*ResolvedResultRef, string, error) {
-	referencedPipelineTask := pipelineState.ToMap()[resultRef.PipelineTask]
-	if referencedPipelineTask == nil {
-		return nil, resultRef.PipelineTask, fmt.Errorf("could not find task %q referenced by result", resultRef.PipelineTask)
-	}
-	if !referencedPipelineTask.isSuccessful() && !referencedPipelineTask.isFailure() {
-		return nil, resultRef.PipelineTask, fmt.Errorf("task %q referenced by result was not finished", referencedPipelineTask.PipelineTask.Name)
+func resolveCustomResultRef(customRuns []*v1beta1.CustomRun, resultRef *v1.ResultRef) (*ResolvedResultRef, error) {
```

Comment:
It looks like the refactor here attempts to separate resolveCustomResultRef out from previously the first condition check at L140. Do we not need the check the following any longer or is this indicating customRuns would also support matrixed referenced Tasks IIUC? 
```
		if len(referencedPipelineTask.CustomRuns) != 1 {
			return nil, resultRef.PipelineTask, fmt.Errorf("referenced tasks can only have length of 1 since a matrixed task does not support producing results, but was length %d", len(referencedPipelineTask.TaskRuns))
		}

```

Reply:
Custom Runs would also support referenced Matrixed tasks. Added an example unit test: https://github.com/tektoncd/pipeline/blob/3e499fe5f2b90bccba1ac7d8a2d3db92d7cbb1bb/pkg/reconciler/pipelinerun/pipelinerun_test.go#L13558